Add Exception Summaries

msagarpatel opened the pull request:
This PR adds the ability to summarize the current exception. This is especially helpful in cases where there are large exceptions with multiple TaskFailedExceptions and CompositeExceptions wrapping multiple exceptions.

The interface is simple:

```Julia
    summarize_current_exceptions(io::IO=Base.stderr)
```

## Examples:

```julia
julia> begin
           try
               try
                   @sync begin
                       Threads.@spawn try
                           @assert false
                       catch
                           @assert 1 != 1
                       end
                       Threads.@spawn @assert 2 != 2
                   end
               catch
                   @assert 3 != 3
               end
           catch
               summarize_current_exceptions()
           end
       end
```
```
=== EXCEPTION SUMMARY ===

CompositeException (2 tasks):
 1. AssertionError: false
     [1] macro expansion
       @ ./REPL[14]:6 [inlined]

    which caused:
    AssertionError: 1 != 1
     [1] macro expansion
       @ ./REPL[14]:8 [inlined]
 --
 2. AssertionError: 2 != 2
     [1] (::var"#10#12")()
       @ Main ./threadingconstructs.jl:258

which caused:
AssertionError: 3 != 3
 [1] top-level scope
   @ REPL[14]:13
```

## TODOs

- [ ] Test this in a larger codebase that can generate truly HUGE exceptions
- [x] Unit tests (WIP already working on this)
- [ ] Seen set, for deduplication
- [ ] Colorize the exception, not the "caused by"
- [x] Add examples to this description
- [ ] Add ability to pass arbitrary exception rather than only allow current exception

msagarpatel committed:
checkpoint from last night

msagarpatel committed:
add tests that have already been created

NHDaly commented on src/exception_summary.jl:
```diff
@@ -0,0 +1,98 @@
+#     Pros: less noise
+#     Cons: possibly hiding intermediate exceptions that might have been helpful to see.
+
+const TITLE = "=== THE NEEDLE ==="
```
😁 as much as i do agree this was fun, i'm going to change this to something that is more obvious when just reading the log message 😏  ❤️ haha sorry
```suggestion
const TITLE = "=== EXCEPTION SUMMARY ==="
```

msagarpatel replied:
😭

NHDaly committed:
Update src/exception_summary.jl

NHDaly commented on src/exception_summary.jl:
```diff
@@ -0,0 +1,98 @@
+    println(io)
+
+    # Print the source line number of the where the exception occurred.
+    bt = Base.process_backtrace(stack)
+    # borrowed from julia/base/errorshow.jl
+    (frame, n) = bt[1]
```
In the compilation profiling project, we discovered that just calling `Base.process_backtrace()` can be really slow in prod for us, since our binary is so big!

So i think it's better to grab just the first frame from the backtrace _first_, and then either leave all the inlined names, or still filter it one more time:
```suggestion
    bt = Base.process_backtrace(stack[1:1])
    # borrowed from julia/base/errorshow.jl
    (frame, n) = bt[1]
```

NHDaly replied:
🤔 but i also seem to remember that we tried something like this but it didn't always work.. 🤔 lemme play with it locally

NHDaly replied:
Hokay, i think i got something that should work, here: e1fa423

msagarpatel replied:
👍 that commit LGTM. Thanks!

NHDaly committed:
Perf improvement printing stacktrace (& small cleanup)

Filter the stacktrace _before_ processing it all, which can improve perf
quite a bit.

Before:
```julia
julia> try
           @assert false
       catch
           @time summarize_current_exceptions()
       end
=== EXCEPTION SUMMARY ===
AssertionError: false
 [1] top-level scope
  @ ./REPL[95]:2
  0.004641 seconds (484 allocations: 27.102 KiB)
```

After:
```julia
julia> try
           @assert false
       catch
           @time summarize_current_exceptions()
       end
=== EXCEPTION SUMMARY ===
AssertionError: false
 [1] top-level scope
  @ ./REPL[95]:2
  0.001720 seconds (228 allocations: 12.766 KiB)
```

NHDaly committed:
Fix tests for julia 1.0-1.3

NHDaly commented on test/exception_summary.jl:
```diff
@@ -0,0 +1,142 @@
+using ExceptionUnwrapping: summarize_current_exceptions, TITLE, INDENT_LENGTH, SEPARATOR
+
+function get_current_exception_string()
+    io = IOBuffer()
+    summarize_current_exceptions(io)
+    str = String(take!(io))
```
Up to you, but this is shorthand for the same thing you have here:
```suggestion
    str = sprint(summarize_current_exceptions)
```

msagarpatel replied:
Oh cool, I learned something new 😁 I added `str = ` your suggestion, and accepted it.

NHDaly committed:
Numbered Composite items + Support for multiline exceptions

- Number the items in a CompositeException when printing.
- Support multiline exception strings, still indented properly on each
  of their multiple lines.
- Add a test for these situations.

For example:
```julia
=== EXCEPTION SUMMARY ===

CompositeException (length 2):
 1. MultiLineException(
        0
    )
     [1] throw_multiline(x::Int64)
       @ Main foo.jl:23

    which caused:
    MultiLineException(
        1
    )
     [1] throw_multiline(x::Int64)
       @ Main foo.jl:23

 2. MultiLineException(
        2
    )
     [1] throw_multiline(x::Int64)
       @ Main foo.jl:23

which caused:
MultiLineException(
    3
)
 [1] throw_multiline(x::Int64)
   @ Main foo.jl:23
```

NHDaly committed:
Simplify test logic

NHDaly committed:
Change CompositeException "(length 2)" to "(2 tasks)"

This hopefully makes it clearer what's going on, for a reader who
doesn't already know what a CompositeException is for.

NHDaly committed:
make the Composite sep more subtle

now that the items are numbered

NHDaly committed:
Make test strings more explicit

NHDaly committed:
Only enable the exception_summary for julia 1.7+

NHDaly committed:
CI

NHDaly commented:
@msagarpatel:
Regarding the TODOs:

- Test this in a larger codebase that can generate truly HUGE exceptions
    - Good idea, but we could probably merge this PR as-is, and then iterate!
    - No need to let perfection be the enemy of the good.
- Unit tests (WIP already working on this)
    - I think the tests we have now seem decent! What else would you want to add?
- Seen set, for deduplication
    - I think we can do this in a follow-up. No need for it now.
- Colorize the exception, not the "caused by"
    - Meh.. I don't think this is easy, and i don't think it's worth doing.
    - Base doesn't color the exception names, and we're reusing their printing, so i don't think this is feasible, nor important.
- Add examples to this description
    - sure, sounds nice. I can copy in one of the examples I added as a test. Is that enough?
- Add ability to pass arbitrary exception rather than only allow current exception
    - Fine to leave this as a follow-up, or never do it. I don't think this really makes
      sense considering that the whole goal here is to print a whole exception stack.
    - With just a single exception, you could just call Base.display_error().

NHDaly reviewed (approved):
@msagarpatel: I think this PR LGTM! 

I'll leave it open in case there's anything else you want to address, but this looks really great. :)

I made a few more improvements today, and I think this is ready to use! We should plug it in to RAICode, since it's not helping anyone just sitting here. 😊 We can always make more changes later, but i think this would already be really, really helpful.

msagarpatel committed:
use sprint

Co-authored-by: Nathan Daly <NHDaly@gmail.com>

msagarpatel commented:
Thanks @NHDaly! I'm still going through your comments and changes. Gonna check out for the night, but I'm aiming to wrap this up tomorrow.
